src/features/device-provisioning/hooks/cache-invalidation.ts

```diff
@@ -0,0 +1,45 @@
+import { sbvrUtils, permissions, hooks } from '@balena/pinejs';
+import {
+	checkApiKeyExistsStore,
+	checkDeviceUuidExistsStore,
+} from '../gracefully-deny-conflicting-registrations';
+import type { createMultiLevelStore } from '../../../infra/cache';
+
+const setupCacheInvalidation = <K extends string>(
+	model: string,
+	resource: string,
+	keyProperty: K,
+	cache: ReturnType<typeof createMultiLevelStore<any>>,
+) => {
+	hooks.addPureHook('DELETE', model, resource, {
+		PRERUN: async (args) => {
+			const { api, tx } = args;
+			const affectedIds = await sbvrUtils.getAffectedIds(args);
+			if (affectedIds.length === 0) {
+				return;
+			}
+			const affectedItems = (await api.get({
+				resource,
+				passthrough: { req: permissions.root, tx },
+				options: {
+					$select: keyProperty,
+					$filter: {
+						id: { $in: affectedIds },
+					},
+				},
+			})) as Array<{ [key in K]: string }>;
+
+			// Invalidate the caches only when the tx is committed
+			tx.on('end', async () => {
+				for (const affectedItem of affectedItems) {
+					// Run in the background as this is not a reason to fail the request
+					cache.delete(affectedItem[keyProperty]);
```

Comment:
If the delete fails then it'll be an unhandled promise rejection and cause a bunch of warnings